File: dagger2demo/src/main/java/com/example/dagger2demo/MainActivity.java
package com.example.dagger2demo;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;

import javax.inject.Inject;

class MainActivity extends AppCompatActivity {

    /**
     * 标明需要注入的对象
     */
    @Inject
    Person person;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        TextView txtPerson = findViewById(R.id.txt_main_person);

        /**
         * 构造桥梁对象
         */
        MainComponent component = DaggerMainComponent.builder().mainModule(new MainModule()).build();

        /**
         * 注入
         */
        component.inject(this);

        person.name = "liuhe";
        person.age = 18;

        txtPerson.setText(String.format("姓名：%s，年龄：%d", person.name, person.age));
    }
}

File: aidlclient/src/main/java/com/liuhe/aidlclient/MainActivity.kt
package com.liuhe.aidlclient

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.support.v7.app.AppCompatActivity
import com.liuhe.aidlservice.IRemoteInterface
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    init {
        println("init$companionField")
    }

    companion object {
        var companionField = "haha"
    }

    var iService: IRemoteInterface? = null
    private var childHandler: Handler? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        btn_main_bind.setOnClickListener({
            var intent = Intent()
            intent.action = "com.liuhe.aidlservice"
            /*
             服务端包名
            有些时候我们使用Service的时需要采用隐私启动的方式，但是Android 5.0一出来后，其中有个特性就是Service Intentmust be explitict，
            也就是说从Lollipop开始，service服务必须采用显示方式启动。
             */
            intent.`package` = "com.liuhe.aidlservice"
            bindService(intent, MyConn(), Context.BIND_AUTO_CREATE)
        })

        btn_main_call.setOnClickListener({
            iService?.doSomething("我是Aidl Client")
        })
    }

    inner class MyConn : ServiceConnection {
        override fun onServiceDisconnected(name: ComponentName?) {
        }

        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            iService = IRemoteInterface.Stub.asInterface(service)
        }
    }



    /**
     * 子线程之间的通信
     */
    private fun connChildThread() {
        Thread(Runnable {
            Looper.prepare()
            childHandler = @SuppressLint("HandlerLeak")
            object : Handler() {
                override fun handleMessage(msg: Message) {
                    super.handleMessage(msg)
                    System.out.println("这个消息是从-->>" + msg.obj + "过来的，在" + "btn的子线程当中" + "中执行的")
                }
            }
            Looper.loop()//开始轮循
        }).start()

        Thread(Runnable {
            val msg = childHandler?.obtainMessage()
            msg?.obj = "btn2当中子线程"
            childHandler?.sendMessage(msg)
        }).start()
    }
}
